### DISCUSSION TITLE
`hx --grammar build` fails with Groovy tree-sitter grammar on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by holtkampjs** [helix-editor/helix#9228] [Troubleshooting] [answered]
# Description

I'm attempting to set up a tree-sitter grammar for Groovy: [tree-sitter-groovy](https://github.com/Decodetalkers/tree-sitter-groovy).

## Setup

The addition I made to my `languages.toml` is as follows:

```toml
[[language]]
name = "groovy"
scope = "source.groovy"
roots = []
file-types = ["groovy"]

[[grammar]]
name = "groovy"
source = { git = "https://github.com/Decodetalkers/tree-sitter-groovy", rev = "4c4c13eb3121f6bdc76d3ead9a1083ed08e70177" }
```

## Execution

The fetch succeeded.

```fish
hx --grammar fetch
```

The build command failed with the following output:

```fish
hx --grammar build
```

```txt
Building 137 grammars
136 grammars already built
Failure 1/1: groovy Failed to execute C/C++ compiler
Error: 1 grammars failed to build
```

## Debugging

I searched the issues in this repository: [Issues](https://github.com/helix-editor/helix/issues?q=%22Failed+to+execute+C%2FC%2B%2B+compiler%22).

The general advice was to check that both `gcc` and `g++` are installed.

```fish
$ gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ g++ --version
g++ (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR P
```

The other advice were to check/set the `CC` environment variable. Mine was set to nothing, so I set it to the path of the `gcc` executable and still received the same error when building.

```fish
$ echo $CC

$ which gcc
/usr/bin/gcc
$ set -Ux CC /usr/bin/gcc
$ echo $CC
/usr/bin/gcc
$ hx --grammar build
Building 137 grammars
136 grammars already built
Failure 1/1: groovy Failed to execute C/C++ compiler
Error: 1 grammars failed to build
```

I realized the fact that I'm working in a [`fish` shell](https://fishshell.com/) could be causing issues. So, I opened a `bash` shell and saw the same output for both `echo $CC` and `hx --build grammar`.

# System Infomation

Operating System: Ubuntu 20.04
Shell: Fish
Helix: `helix 23.10 (f6021dd0)`

```bash
$ uname -a
Linux wdxd3ks5m3 5.15.0-91-generic #101~20.04.1-Ubuntu SMP Thu Nov 16 14:22:28 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

## Notes

I really love using Helix as my main editor. The rough edges I've encountered have been dealing with languages that are not supported out of the box (`groovy` and `powershell`). Any help would be greatly appreciated.

**Comment by the-mikedavis** [comment, marked as answer]
I can reproduce the build failure. That groovy parser doesn't check in the C files under `src/` into git that we compile, so the parsing step fails

**Reply by holtkampjs** [reply]
Looking under my `~/.config/helix/runtime/grammars/sources/groovy/` directory, I can see that there isn't a `src/` folder. Are there any workarounds? 

- Should I fork the repo, build it and check it into git with the contents of the `src/` directory intact?
- Build it in the `~/.config/helix/runtime/grammars/sources/groovy/` directory?
- Any other suggestions?

**Reply by kirawi** [reply]
You can do `source.path` to point to a local location

**Reply by apraga** [reply]
Cloning the repository, running `tree-sitter generate` generate the `src`directiory.
Then `hx --grammar buid` succeed with the following configuration.

```
[[language]]
name = "groovy"
scope = "source.groovy"
roots = []
file-types = ["nf", "nextflow", "groovy"]

[[grammar]]
name = "groovy"
source = { path = "/home/alex/code/tree-sitter-groovy" }
```

However, there is no syntax highlighting in a .groovy file

**Reply by kirawi** [reply]
Did you copy the queries to your runtime folder? https://docs.helix-editor.com/guides/adding_languages.html

**Reply by apraga** [reply]
@kirawi There does not seem to be a `queries` folder so that explains the absence of highlighting !

**Reply by Zoybean** [reply]
I'm having the same issue on windows 10 (without WSL), but with near-identical groovy language configuration.
Just filling in the parts that I needed that weren't documented in the accepted answer:
- ran `cargo install tree-sitter-cli` to get the [`tree-sitter` shell command](https://github.com/tree-sitter/tree-sitter/blob/master/cli/README.md)
- `tree-sitter generate` complained about [`node`](https://nodejs.org/en/download) not being installed, so I installed that
  - node installation failed to install some part of the C/C++ compiler tools, likely because they're already been installed. I ignored the error and rebooted.

After doing the above, `tree-sitter generate` successfully runs, `hx --grammar build` succeeds for all languages, but `hx --health groovy` gives the same results as before building the grammar:
```
> hx --health groovy
Configured language servers: None
Configured debug adapter: None
Highlight queries: ✘
Textobject queries: ✘
Indent queries: ✘
```

~~I have a `runtime/queries` directory, but there's no `runtime/queries/groovy`. Have I missed a step to generate the queries?~~
Update: I reread the instructions - I misunderstood the step to copy the queries. As `Decodetalkers/tree-sitter-groovy` doesn't include its own queries, I was a bit stuck, but the repo's one issue links to [`murtaza64/tree-sitter-groovy`](https://github.com/murtaza64/tree-sitter-groovy) which builds without issue and does include queries. After that, running `cp -r runtime\grammars\sources\groovy\queries\ runtime\queries\groovy` resolved the issue.
After going through all that, it seems the real solution for me was to just switch from `Decodetalkers/tree-sitter-groovy` to `murtaza64/tree-sitter-groovy`. All the other steps were just workarounds for the former that weren't needed for the latter.
